desc/compute/_stability.py

```diff
@@ -232,3 +232,399 @@ def _magnetic_well(params, transforms, profiles, data, **kwargs):
+
+
+@register_compute_fun(
+    name="ideal ball lambda",
+    label="\\lambda_{\\mathrm{ballooning}}=\\gamma^2",
+    units="~",
+    units_long="None",
+    description="Normalized squared ideal ballooning growth rate, "
+    "requires data along a field line",
+    dim=1,
+    params=["Psi"],
+    transforms={"grid": []},
+    profiles=[],
+    coordinates="rtz",
+    data=[
+        "a",
+        "g^aa",
+        "g^ra",
+        "g^rr",
+        "cvdrift",
+        "cvdrift0",
+        "|B|",
+        "B^zeta",
+        "p_r",
+        "iota",
+        "shear",
+        "psi",
+        "psi_r",
+        "rho",
+    ],
+    source_grid_requirement={"coordinates": "raz", "is_meshgrid": True},
+    zeta0="array: points of vanishing integrated local shear to scan over. "
+    "Default 15 points linearly spaced in [-π/2,π/2]",
+)
+def _ideal_ballooning_gamma2(params, transforms, profiles, data, **kwargs):
+    """
+    Ideal-ballooning growth rate finder.
+
+    This function uses a finite-difference method
+    to calculate the maximum growth rate against the
+    infinite-n ideal ballooning mode. The equation being solved is
+
+    d/dζ(g dX/dζ) + c X = λ f X, g, f > 0
+
+    where
+
+    𝛋 = b ⋅∇ b
+    g = a_N^3 * B_N * (b ⋅∇ζ) * (dψ_N/dρ)² * |∇α|², / B,
+    c = a_N^3 * B_N * (1/ b ⋅∇ζ) * (dψ_N/dρ)² * dp/dψ * (b × 𝛋) ⋅|∇α|/ B**2,
+    f = a_N * B_N^3 * (dψ_N/dρ)² * |∇α|² / B^3 * (1/ b ⋅∇ζ) ,
+
+    are needed along a field line to solve the ballooning equation once and
+    find
+
+    λ = a_N^2 / v_A^2 * γ²,
+
+    where
+
+    v_A = B_N /sqrt(mu_0 * n0 * M) is the Alfven speed, and
+    ψ_N = ψ/ψ_b is the normalized toroidal flux, and
+    ψ_b = 0.5*(B_N * a_N**2) is the total enclosed toroidal flux.
+
+    To obtain the parameters g, c, and f, we need a set of parameters
+    provided in the list ``data`` above. Here's a description of
+    these parameters:
+
+    - a: minor radius of the device
+    - g^aa: |grad alpha|^2, field line bending term
+    - g^ra: (grad alpha dot grad rho) integrated local shear
+    - g^rr: |grad rho|^2 flux expansion term
+    - cvdrift: geometric factor of the curvature drift
+    - cvdrift0: geometric factor of curvature drift 2
+    - |B|: magnitude of the magnetic field
+    - B^zeta:  B dot grad zeta
+    - p_r: dp/drho, pressure gradient
+    - phi: coordinate describing the position in the toroidal angle
+    along a field line
+    """
+    source_grid = transforms["grid"].source_grid
+    # Vectorize in rho later
+    rho = source_grid.meshgrid_reshape(data["rho"], "arz")
+
+    psi_b = params["Psi"] / (2 * jnp.pi)
+    a_N = data["a"]
+    B_N = 2 * psi_b / a_N**2
+
+    zeta0 = kwargs.get("zeta0", jnp.linspace(-0.5 * jnp.pi, 0.5 * jnp.pi, 15))
+    N_zeta0 = len(zeta0)
+
+    # This would fail with rho vectorization
+    iota = jnp.mean(data["iota"])
+    shear = jnp.mean(data["shear"])
+    psi = jnp.mean(data["psi"])
+    sign_psi = jnp.sign(psi)
+    sign_iota = jnp.sign(iota)
```

Comment:
Aren't these averaging over all surfaces? I don't think that's what we want, but maybe I'm misunderstanding the grid being used

Reply:
Wait, does this only work assuming the grid is a single flux surface? Do we have a `source_grid_requirement` for that?

Reply:
Yes, only a single flux surface right now. source_grid works fine for Nrho=1 if that's what you are asking.